runtimebp/cpu.go

```diff
@@ -127,11 +134,108 @@ func defaultMaxProcsFormula(n float64) int {
+	// the environment variable is set
+	v, _ := os.LookupEnv("BASEPLATE_GOMAXPROCSMULT")
+	intVal, err := strconv.Atoi(v)
+	if err != nil {
+		fmt.Fprintf(
+			os.Stderr,
+			"runtimebp.fetchMaxProcsMult: GOMAXPROCSMULT is set to %q, which is invalid, falling back to %d to ensure parallelism: %v",
+			v,
+			defaultMultiplier,
+			err,
+		)
+		intVal = defaultMultiplier
+	}
+	if intVal <= 0 {
+		intVal = 1
+	}
+	return intVal
+}
+
+// fetchCPURequest checks for the magic BASEPLATE_CPU_REQUEST variable set as a fallback
+// by Infrared.  If this is set, we use it and the multiplier to set our
+// GOMAXPROCS as a fallback.
+func fetchCPURequest() (int, bool) {
+	var req float64
+	v, ok := os.LookupEnv("BASEPLATE_CPU_REQUEST")
+	if !ok {
+		return 0, false
+	}
+
+	iReq, err := strconv.Atoi(v)
+	if err != nil {
+		// This case will probably never be hit, but I'm keeping it just in case.
+		// In k8s 1.21 and newer, when you present the CPU request via the Downward API
+		// k8s automatically rounds it up to the nearest integer unit.  That's how
+		// we plan to set this variable, but just in case it might be worth doing this
+		// check in case they ever break that.  There's no official contract for that.
+		match := millicoreRegexp.FindStringSubmatch(v)
+		if match == nil {
+			fmt.Fprintf(
+				os.Stderr,
+				"runtimebp.fetchCPURequest: BASEPLATE_CPU_REQUEST is set to %q, which is invalid, ignoring.",
+				v,
+			)
+			req = 1
+		} else {
+			m, err := strconv.Atoi(match[1])
+			if err == nil {
+				req = float64(m) / 1000
+			} else {
+				fmt.Fprintf(
+					os.Stderr,
+					"runtimebp.fetchCPURequest: BASEPLATE_CPU_REQUEST is set and appears to be a millicore value (%v), but not an integer: %v",
+					req,
+					err,
+				)
+				req = 1
+			}
+		}
+	} else {
+		req = float64(iReq)
+	}
+	return scaledMaxProcsFormula(req), true
+}
+
+// GOMAXPROCS sets runtime.GOMAXPROCS with the relevant formula,
 // in bound of [min, max].
 //
-// Currently the default formula is NumCPU() rounding up.
+// Start by checking if limits are set at all.
+// If they aren't, fall back to the CPU request for the container
+// as provided by Infrared, with multiplier, if it exists.
+// If limits are set, check for a multiplier and multiply by it.
+// If limits are set with no multiplier, use a default multiplier.
+// If no limits are set and no request is provided by Infrared,
+// just use the default multiplier outright as our GOMAXPROCS value.
 func GOMAXPROCS(min, max int) (oldVal, newVal int) {
+	var limitSet error
+	buf := make([]byte, 1024)
+	_, limitSet = numCPUCgroupsV2(buf)
+	if limitSet == nil {
+		_, limitSet = numCPUCgroupsV1(buf)
+		if limitSet == nil {
+			if req, ok := fetchCPURequest(); ok {
+				oldVal = runtime.GOMAXPROCS(req)
+				return oldVal, req
+			}
+		}
+	}
+	if _, ok := os.LookupEnv("BASEPLATE_GOMAXPROCSMULT"); ok {
+		return GOMAXPROCSwithFormula(min, max, scaledMaxProcsFormula)
```

Comment:
since we already read and multiplied the multiplier in `fetchCPURequest`, do we really need this fallback? what do we multiply it with without the requests env? 1?

if that's the case, it's better to call the formula directly (vs. using `GOMAXPROCSwithFormula`, which would call `NumCPU` and try to read the cgroup limits and print error logs again).

we want to first refactor the part of `GOMAXPROCSwithFormula` to a helper function:

```
func setGOMAXPROCSwithBoundary(n, min, max int) (oldVal, newVal int) {
	newVal = boundNtoMinMax(n, min, max)
	oldVal = runtime.GOMAXPROCS(newVal)
	return oldVal, newVal
}
```

then use that directly:
```suggestion
		return setGOMAXPROCSwithBoundary(scaledMaxProcsFormula(1), min, max)
```